src/middlewared/middlewared/plugins/idmap.py

```diff
@@ -0,0 +1,1103 @@
+            raise CallError(f'Deletion of system idmap domain [{entry["name"]}] is not permitted.', errno.EPERM)
+        await self.middleware.call("datastore.delete", self._config.datastore, id)
+
+    @accepts(Int("id"))
+    async def run(self, id):
+        data = await self._get_instance(id)
+
+    @private
+    async def _validate(self, data):
+        verrors = ValidationErrors()
+        if data['id'] <= dstype['DS_TYPE_DEFAULT_DOMAIN'].value:
+            verrors.append(f'Modification of system idmap domain [{data["name"]}] is not permitted.')
+        return verrors
+
+
+class IdmapDomainBackendService(CRUDService):
+    class Config:
+        datastore = 'directoryservice.idmap_domaintobackend'
+        datastore_prefix = 'idmap_dtb_'
+        namespace = 'idmap.domaintobackend'
+
+
+    @accepts(
+        Dict(
+            'idmap_domaintobackend_create',
+            Dict(
+                 'domain',
+                 Int('id'),
+                 Str('idmap_domain_name'),
+                 Str('idmap_domain_dns_domain_name'),
+            ),
+            Any('idmap_backend'),
+            register=True
+        )
+    )
+    async def do_create(self, data):
+        """
+        Enforce referential integrity during domain-to-backend creation
+        by also creating a corresponding entry in the correct idmap backend table.
+        """
+        verrors = ValidationErrors()
+        if verrors:
+            raise verrors
+
+        data = await self.middleware.call('idmap.common_backend_compress', data)
+        self.logger.debug(data)
+        backend_entry_is_present = False
+        idmap_data = await self.middleware.call(f'idmap.{data["idmap_backend"]}.query')
+        for i in idmap_data:
+            if not i['domain']:
+                continue
+            if i['domain']['idmap_domain_name'] == data['domain']['idmap_domain_name']:
+                backend_entry_is_present
+                break
+
+        if not backend_entry_is_present:
+            next_idmap_range = await self.get_next_idmap_range()
+            new_idmap = await self.middleware.call(f'idmap.{data["idmap_backend"]}.create', {
+                'domain': {'id': data['domain']['id']},
+                'range_low': next_idmap_range[0],
+                'range_high': next_idmap_range[1]
+            })
+        data["id"] = await self.middleware.call(
+            "datastore.insert", self._config.datastore, data,
+            {
+                "prefix": self._config.datastore_prefix
+            },
+        )
+        return await self._get_instance(data['id'])
+
+    @accepts(
+        Int('id', required=True),
+        Patch(
+            "idmap_domaintobackend_create",
+            "idmap_domaintobackend_update",
+            ("attr", {"update": True})
+        )
+    )
+    async def do_update(self, id, data):
+        old = await self._get_instance(id)
+        new = old.copy()
+        new.update(data)
+        verrors = ValidationErrors()
+
```

Comment:
Is there any validation missing? if not we should remove verrors

Reply:
Removed various unnecessary validation stubs. I will add validation on an as-needed basis.